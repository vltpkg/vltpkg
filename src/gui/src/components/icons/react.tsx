export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
