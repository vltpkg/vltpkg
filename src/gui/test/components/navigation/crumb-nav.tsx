import { vi, test, expect, afterEach, describe } from 'vitest'
import { render, cleanup, fireEvent } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.ts'
import {
  getBreadcrumbs,
  CrumbNav,
} from '@/components/navigation/crumb-nav.tsx'

vi.mock('@/components/query-bar/query-token.tsx', () => ({
  QueryToken: 'gui-query-token',
}))

vi.mock('lucide-react', () => ({
  ChevronRight: 'gui-chevron-right-icon',
  Ellipsis: 'gui-ellipsis-icon',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('CrumbNav component', () => {
  test('does not render without breadcrumbs', () => {
    const Container = () => {
      return <CrumbNav breadcrumbs={undefined} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toBe('')
  })

  test('renders single breadcrumb item', () => {
    const breadcrumbs = getBreadcrumbs('#express')

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders multiple breadcrumb items with separators', () => {
    const breadcrumbs = getBreadcrumbs(':root > #express > #lodash')

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders breadcrumb with semver selector', () => {
    const breadcrumbs = getBreadcrumbs('#express:semver(^4.0.0)')

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders complex breadcrumb path', () => {
    const breadcrumbs = getBreadcrumbs(
      ':project > #express:v(4) > #lodash > #underscore',
    )

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders workspace breadcrumb', () => {
    const breadcrumbs = getBreadcrumbs(':workspace > #utils')

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders with custom className', () => {
    const breadcrumbs = getBreadcrumbs('#express > #lodash')

    const Container = () => {
      return (
        <CrumbNav
          breadcrumbs={breadcrumbs}
          className="custom-nav-class"
        />
      )
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('renders truncated breadcrumbs when more than 5 items', () => {
    // Create a long breadcrumb path with 10 items
    const longQuery =
      ':root > #a > #b > #c > #d > #e > #f > #g > #h > #i'
    const breadcrumbs = getBreadcrumbs(longQuery)

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('shows ellipsis button in truncated view', () => {
    const longQuery =
      ':root > #a > #b > #c > #d > #e > #f > #g > #h > #i'
    const breadcrumbs = getBreadcrumbs(longQuery)

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container, getByLabelText } = render(<Container />)

    // Should show ellipsis button
    const ellipsisButton = getByLabelText('Show hidden breadcrumbs')
    expect(ellipsisButton).toBeDefined()
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('expands breadcrumbs when ellipsis button is clicked', () => {
    const longQuery =
      ':root > #a > #b > #c > #d > #e > #f > #g > #h > #i'
    const breadcrumbs = getBreadcrumbs(longQuery)

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container, getByLabelText } = render(<Container />)

    // Click ellipsis button to expand
    const ellipsisButton = getByLabelText('Show hidden breadcrumbs')
    fireEvent.click(ellipsisButton)

    expect(container.innerHTML).toMatchSnapshot()
  })

  test('does not truncate when 5 or fewer breadcrumbs', () => {
    // Create exactly 5 breadcrumbs
    const mediumQuery = ':root > #a > #b > #c > #d'
    const breadcrumbs = getBreadcrumbs(mediumQuery)

    const Container = () => {
      return <CrumbNav breadcrumbs={breadcrumbs} />
    }

    const { container, queryByLabelText } = render(<Container />)

    // Should not show ellipsis button
    const ellipsisButton = queryByLabelText('Show hidden breadcrumbs')
    expect(ellipsisButton).toBeNull()
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('handles undefined breadcrumbs gracefully', () => {
    const Container = () => {
      return <CrumbNav breadcrumbs={undefined} />
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toBe('')
  })
})

describe('getBreadcrumbs function', () => {
  test('returns undefined for root query', () => {
    const breadcrumbs = getBreadcrumbs(':root')
    expect(breadcrumbs).toBeUndefined()
  })

  test('returns breadcrumb for simple id selector', () => {
    const breadcrumbs = getBreadcrumbs('#express')
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(true)
    expect(breadcrumbs?.first.value).toBe('#express')
    expect(breadcrumbs?.first.type).toBe('id')
  })

  test('returns breadcrumb for complex query', () => {
    const breadcrumbs = getBreadcrumbs(':root > #express > #lodash')
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(false)
    expect(breadcrumbs?.first.value).toBe(':root')
    expect(breadcrumbs?.last.value).toBe('#lodash')
  })

  test('returns breadcrumb for semver query', () => {
    const breadcrumbs = getBreadcrumbs('#express:semver(^4.0.0)')
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(true)
    expect(breadcrumbs?.first.value).toBe('#express:semver(^4.0.0)')
    expect(breadcrumbs?.first.type).toBe('id')
  })

  test('returns undefined for invalid query', () => {
    const breadcrumbs = getBreadcrumbs('[invalid-selector]')
    expect(breadcrumbs).toBeUndefined()
  })

  test('returns breadcrumb for workspace selector', () => {
    const breadcrumbs = getBreadcrumbs(':workspace')
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(true)
    expect(breadcrumbs?.first.value).toBe(':workspace')
    expect(breadcrumbs?.first.type).toBe('pseudo')
    expect(breadcrumbs?.first.importer).toBe(true)
  })

  test('returns breadcrumb for project selector', () => {
    const breadcrumbs = getBreadcrumbs(':project')
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(true)
    expect(breadcrumbs?.first.value).toBe(':project')
    expect(breadcrumbs?.first.type).toBe('pseudo')
    expect(breadcrumbs?.first.importer).toBe(true)
  })

  test('returns breadcrumb for long query', () => {
    const longQuery =
      ':root > #a > #b > #c > #d > #e > #f > #g > #h > #i'
    const breadcrumbs = getBreadcrumbs(longQuery)
    expect(breadcrumbs).toBeDefined()
    expect(breadcrumbs?.single).toBe(false)

    // Count breadcrumbs
    const count = breadcrumbs ? [...breadcrumbs].length : 0
    expect(count).toBe(10)
  })
})
