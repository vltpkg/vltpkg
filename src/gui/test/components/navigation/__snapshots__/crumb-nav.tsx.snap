// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CrumbNav component > does not truncate when 5 or fewer breadcrumbs 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :root
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #a
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #b
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #c
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #d
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > expands breadcrumbs when ellipsis button is clicked 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :root
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #a
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #b
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #c
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #d
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #e
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #f
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #g
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #h
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #i
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders breadcrumb with semver selector 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #express
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders complex breadcrumb path 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :project
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #express
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #lodash
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #underscore
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders multiple breadcrumb items with separators 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :root
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #express
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #lodash
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders single breadcrumb item 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #express
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders truncated breadcrumbs when more than 5 items 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :root
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button
    class="inline-flex items-center justify-center rounded px-1 py-0.5 text-muted-foreground opacity-75 transition-opacity hover:opacity-100 hover:bg-muted/50"
    aria-label="Show hidden breadcrumbs"
  >
    <gui-ellipsis-icon size="14">
    </gui-ellipsis-icon>
  </button>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #g
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #h
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #i
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders with custom className 1`] = `

<nav class="flex flex-row items-center gap-1 custom-nav-class">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #express
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #lodash
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > renders workspace breadcrumb 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :workspace
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #utils
    </gui-query-token>
  </button>
</nav>

`;

exports[`CrumbNav component > shows ellipsis button in truncated view 1`] = `

<nav class="flex flex-row items-center gap-1">
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="pseudo"
    >
      :root
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button
    class="inline-flex items-center justify-center rounded px-1 py-0.5 text-muted-foreground opacity-75 transition-opacity hover:opacity-100 hover:bg-muted/50"
    aria-label="Show hidden breadcrumbs"
  >
    <gui-ellipsis-icon size="14">
    </gui-ellipsis-icon>
  </button>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #g
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5 duration-250 opacity-75 transition-opacity hover:opacity-100"
      variant="id"
    >
      #h
    </gui-query-token>
  </button>
  <gui-chevron-right-icon
    size="14"
    strokewidth="2"
    classname="text-muted-foreground opacity-75"
  >
  </gui-chevron-right-icon>
  <button class="inline-flex">
    <gui-query-token
      classname="px-1.5"
      variant="id"
    >
      #i
    </gui-query-token>
  </button>
</nav>

`;
